src/main/java/teammates/common/util/ActivityLogEntry.java

```diff
@@ -607,25 +372,126 @@ public void highlightKeyStringInMessageInfoHtml() {
+     * All null values (if possible) that are passed into the builder will be ignored.
+     * 
+     * @see {@link ActivityLogEntry}
+     */
+    public static class Builder {
+        // Required parameters
+        private String actionName;
+        private String actionUrl;
+        private long logTime;
+        
+        // Optional parameters - initialized to default values
+        private String actionResponse = Const.ActivityLog.UNKNOWN;
+        private long actionTimeTaken;
+        private String userRole = Const.ActivityLog.UNKNOWN;
+        private String userName = Const.ActivityLog.UNKNOWN;
+        private String userGoogleId = Const.ActivityLog.UNKNOWN;
+        private String userEmail = Const.ActivityLog.UNKNOWN;
+        private String logMessage = Const.ActivityLog.UNKNOWN;
+        private String logId = Const.ActivityLog.UNKNOWN;
+        private boolean isMasqueradeUserRole;
+        
+        public Builder(String name, String url, long time) {
+            actionName = name == null ? Const.ActivityLog.UNKNOWN : name;
+            actionUrl = url == null ? Const.ActivityLog.UNKNOWN : url;
+            logTime = time;
+            
+            actionResponse = actionName;
+        }
+        
+        public Builder withActionResponse(String val) {
+            if (val != null) {
+                actionResponse = val;
+            }
+            return this;
+        }
+        
+        public Builder withUserRole(String val) {
+            if (val != null) {
+                userRole = val;
+            }
+            return this;
+        }
+        
+        public Builder withUserName(String val) {
+            if (val != null) {
+                userName = val;
+            }
+            return this;
+        }
+        
+        public Builder withUserGoogleId(String val) {
+            if (val != null) {
+                userGoogleId = val;
+            }
+            return this;
+        }
+        
+        public Builder withUserEmail(String val) {
+            if (val != null) {
+                userEmail = val;
+            }
+            return this;
+        }
+        
+        public Builder withMasqueradeUserRole(Boolean val) {
+            isMasqueradeUserRole = val;
+            return this;
+        }
+        
+        public Builder withLogId(String val) {
+            if (val != null) {
+                logId = val;
+            }
+            return this;
+        }
+        
+        public Builder withLogMessage(String val) {
+            if (val != null) {
+                logMessage = val;
+            }
+            return this;
+        }
+        
+        public Builder withActionTimeTaken(long val) {
+            actionTimeTaken = val;
+            return this;
+        }
+        
+        public long getLogTime() {
+            return logTime;
+        }
+        
+        public String getActionName() {
```

Comment:
As I don't want to pass the `actionName` everywhere in methods. `actionName` and `logTime` are required fields and cannot be changed once builder is created.